cache/nydus.go

```diff
@@ -0,0 +1,148 @@
+	ocispecs "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+
+	nydusify "github.com/containerd/nydus-snapshotter/pkg/converter"
+)
+
+var nydusAnnotations = []string{nydusify.LayerAnnotationNydusBlob, nydusify.LayerAnnotationNydusBootstrap, nydusify.LayerAnnotationNydusBlobIDs}
+
+func isNydusBlob(desc ocispecs.Descriptor) bool {
+	if desc.Annotations == nil {
+		return false
+	}
+	hasMediaType := desc.MediaType == nydusify.MediaTypeNydusBlob
+	_, hasAnno := desc.Annotations[nydusify.LayerAnnotationNydusBlob]
+	return hasMediaType && hasAnno
+}
+
+// compressNydus compresses a diff tar stream into nydus format (nydus blob + nydus bootstrap).
+func compressNydus(ctx context.Context, comp compression.Config) (compressor, func(context.Context, content.Store) (map[string]string, error)) {
+	return func(dest io.Writer, requiredMediaType string) (io.WriteCloser, error) {
+			return nydusify.Convert(ctx, dest, nydusify.ConvertOption{})
+		}, func(ctx context.Context, cs content.Store) (map[string]string, error) {
+			annotations := map[string]string{
+				// Use this annotation to identify nydus blob layer.
+				nydusify.LayerAnnotationNydusBlob: "true",
+			}
+			return annotations, nil
+		}
+}
+
+// mergeNydus does two steps:
+// 1. Extracts nydus bootstrap from nydus format (nydus blob + nydus bootstrap) for each layer.
+// 2. Merge all nydus bootstraps into a final bootstrap (will as an extra layer).
+// The nydus bootstrap size is very small, so the merge operation is fast.
+func MergeNydus(ctx context.Context, ref ImmutableRef, comp compression.Config, s session.Group) (*ocispecs.Descriptor, error) {
+	iref, ok := ref.(*immutableRef)
+	if !ok {
+		return nil, fmt.Errorf("unsupported ref")
+	}
+	refs := iref.layerChain()
+	if len(refs) == 0 {
+		return nil, fmt.Errorf("refs can't be empty")
+	}
+
+	// Extracts nydus bootstrap from nydus format for each layer.
+	var cm *cacheManager
+	layers := []nydusify.Layer{}
+	blobIDs := []string{}
+	for _, ref := range refs {
+		blobDesc, err := getBlobWithCompressionWithRetry(ctx, ref, comp, s)
+		if err != nil {
+			return nil, errors.Wrapf(err, "get compression blob %q", comp.Type)
+		}
+		ra, err := ref.cm.ContentStore.ReaderAt(ctx, blobDesc)
+		if err != nil {
+			return nil, errors.Wrapf(err, "get reader for compression blob %q", comp.Type)
+		}
+		defer ra.Close()
+		if cm == nil {
+			cm = ref.cm
+		}
+		blobIDs = append(blobIDs, blobDesc.Digest.Hex())
+		layers = append(layers, nydusify.Layer{
+			Digest:   blobDesc.Digest,
+			ReaderAt: ra,
+		})
+	}
+
+	// Merge all nydus bootstraps into a final nydus bootstrap.
+	pr, pw := io.Pipe()
+	go func() {
+		defer pw.Close()
+		if err := nydusify.Merge(ctx, layers, pw, nydusify.MergeOption{
+			WithTar: true,
+		}); err != nil {
+			pw.CloseWithError(errors.Wrapf(err, "merge nydus bootstrap"))
+		}
+	}()
+
+	// Compress final nydus bootstrap to tar.gz and write into content store.
+	chainID := refs[len(refs)-1].getChainID()
```

Comment:
Is the goal here to to have a chainID for all of `refs`? If so, we should just use `iref.getChainID()`. That will be different than the chain ID of the last ref in `refs` when merge and/or diff are being used (which let you slice+join arbitrary layers).

Reply:
Thanks for the reply, fixed that.